### PR TITLE
feat(dummy_traffic_light_publisher): add new node to publish traffic light message instead of Rviz

### PR DESCRIPTION
**Pull request by takam5f2**
## Description

Add `autoware_dummy_traffic_light_publisher` — a dedicated simulator node that publishes dummy traffic light signals for Planning Simulator environments where the traffic light recognition pipeline is not available.

### Background

In the current Planning Simulator workflow, traffic light signals are injected via the `tier4_traffic_light_rviz_plugin` (`TrafficLightPublishPanel`), which publishes directly to `/perception/traffic_light_recognition/traffic_signals`. This procedure is documented in the official Autoware documentation:

- [Traffic light recognition simulation — Set traffic light](https://autowarefoundation.github.io/autoware-documentation/main/demos/planning-sim/traffic-light/#set-traffic-light)

However, an RViz plugin — a GUI visualization tool — should not publish messages directly to topics that the autonomous driving system consumes as sensor/perception input. This creates an implicit and undesirable dependency between the visualization layer and the AD control logic.

rviz_plugin_before_after_v5

This PR introduces a simulator node (`autoware_dummy_traffic_light_publisher`) that sits between external inputs (e.g., the RViz plugin) and the AD system, so that the plugin no longer needs to publish directly to a perception topic.

### Features

- **Standalone mode**: Cycles all traffic lights through Green → Yellow → Red based on the vector map
- **Empty mode**: Publishes empty `TrafficLightGroupArray` (no signal information)
- **Pass-through mode**: Relays external input (e.g., from the RViz plugin) as-is; falls back to the configured mode after a timeout

### Architecture

| Layer | Class | Role |
| --- | --- | --- |
| ROS I/O | `DummyTrafficLightPublisherNode` | Subscriptions (`take()`), timer, publisher, vector map parsing |
| Message assembly | `DummyTrafficLight` | Pass-through judgment, traffic light ID management, `TrafficLightGroupArray` construction |
| Signal generation | `TrafficLightCycle` | Phase transition and `TrafficLightElement` output |

## Related links

**Parent Issue:**

- [feat(dummy_perception_publisher): add publisher for empty traffic light group array #12077](https://github.com/autowarefoundation/autoware_universe/pull/12077)

**Relevant documentation (current behavior to be updated):**

- [Traffic light recognition simulation](https://autowarefoundation.github.io/autoware-documentation/main/demos/planning-sim/traffic-light/)

## How was this PR tested?

- Unit tests for `TrafficLightCycle` (14 tests) and `DummyTrafficLight` (14 tests) — all pass
- Code coverage: `traffic_light_cycle.cpp` 100%, `dummy_traffic_light.cpp` 95% (uncovered lines are closing braces only)
- Verified with Planning Simulator (standalone mode):

  **Ego vehicle passes the intersection on green signal after stopping on red signal:**

standalone_mode_running

  **Ego vehicle is not engaged and traffic lights state changes time by time.**

standalone_mode_stop

## Notes for reviewers

- The node uses `subscription->take()` in the timer callback instead of subscription callbacks
- Logic is fully separated from the ROS Node for testability — Node-level tests are intentionally omitted
- `autoware_launch` integration (adding to `simulator.launch.xml`) and RViz plugin topic redirection will be handled in separate PRs

#### Additions and removals

| Change type | Topic Type | Topic Name | Message Type | Description |
| --- | --- | --- | --- | --- |
| Added | Sub | `~/input/vector_map` | `autoware_map_msgs/msg/LaneletMapBin` | Lanelet2 map to extract traffic light IDs |
| Added | Sub | `~/input/traffic_signals` | `autoware_perception_msgs/msg/TrafficLightGroupArray` | External signals for pass-through |
| Added | Pub | `~/output/traffic_signals` | `autoware_perception_msgs/msg/TrafficLightGroupArray` | Generated or relayed traffic light signals |

#### Modifications

None.

#### Additions and removals

| Change type | Parameter Name | Type | Default Value | Description |
| --- | --- | --- | --- | --- |
| Added | `mode` | `string` | `"empty"` | Operating mode: `"standalone"` or `"empty"` |
| Added | `publish_rate` | `double` | `10.0` | Publishing frequency [Hz] |
| Added | `green_duration` | `double` | `30.0` | Green phase duration [s] |
| Added | `yellow_duration` | `double` | `3.0` | Yellow phase duration [s] |
| Added | `red_duration` | `double` | `30.0` | Red phase duration [s] |
| Added | `passthrough_timeout` | `double` | `1.0` | Timeout before fallback to configured mode [s] |

#### Modifications

None.

## Effects on system behavior

None. This is a new standalone package with no integration into existing launch files yet.

### CODE REVIEW

**Comment by takam5f2** [comment]
@sasakisasaki  
I'd like you to check this PR rather than https://github.com/autowarefoundation/autoware_universe/pull/12077. Best,

File: simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.hpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2026 TIER IV, Inc.
```
Comment (sasakisasaki) [inline simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.hpp:1]: Do we need to make the header public? Same comment for other headers.

File: simulator/autoware_dummy_traffic_light_publisher/include/autoware/dummy_traffic_light_publisher/dummy_traffic_light.hpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+private:
+  autoware_perception_msgs::msg::TrafficLightGroupArray build_standalone_message(
+    const rclcpp::Time & now);
+  autoware_perception_msgs::msg::TrafficLightGroupArray build_empty_message(
+    const rclcpp::Time & now) const;
+
+  Config config_;
+  std::unique_ptr<TrafficLightCycle> cycle_;
+  std::vector<int64_t> traffic_light_ids_;
+  std::optional<autoware_perception_msgs::msg::TrafficLightGroupArray> last_input_;
```
Comment (sasakisasaki) [inline simulator/autoware_dummy_traffic_light_publisher/include/autoware/dummy_traffic_light_publisher/dummy_traffic_light.hpp]: For consistency in the package 👍 
```suggestion
  std::optional<autoware_perception_msgs::msg::TrafficLightGroupArray> last_input_signals_;
```

File: simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+void DummyTrafficLight::set_traffic_light_ids(const std::vector<int64_t> & ids)
+{
+  traffic_light_ids_ = ids;
+}
+
+void DummyTrafficLight::update_input_signals(
+  const autoware_perception_msgs::msg::TrafficLightGroupArray & msg, const rclcpp::Time & now)
+{
+  last_input_ = msg;
```
Comment (sasakisasaki) [inline simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.cpp]: If you apply [my proposal](https://github.com/autowarefoundation/autoware_universe/pull/12456/changes#r3353639575):
```suggestion
  last_input_signals_ = msg;
```

File: simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+autoware_perception_msgs::msg::TrafficLightGroupArray DummyTrafficLight::create_message(
+  const rclcpp::Time & now)
+{
+  if (last_input_ && last_input_time_) {
+    const double elapsed = (now - *last_input_time_).seconds();
+    if (elapsed < config_.passthrough_timeout) {
+      return *last_input_;
+    }
+    last_input_.reset();
+    last_input_time_.reset();
```
Comment (sasakisasaki) [inline simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.cpp]: If you apply [my proposal](https://github.com/autowarefoundation/autoware_universe/pull/12456/changes#r3353639575):
```suggestion
  if (last_input_signals_ && last_input_time_) {
    const double elapsed = (now - *last_input_time_).seconds();
    if (elapsed < config_.passthrough_timeout) {
      return *last_input_signals_;
    }
    last_input_signals_.reset();
    last_input_time_.reset();
```

**Review by sasakisasaki — COMMENTED** [review]
Thank you for your contribution. Quick comment first!

**Comment by mitsudome-r** [comment]
@takam5f2 I think this is a good approach.

> **Standalone mode**: Cycles all traffic lights through Green → Yellow → Red based on the vector map
**Empty mode**: Publishes empty TrafficLightGroupArray (no signal information)
**Pass-through mode**: Relays external input (e.g., from the RViz plugin) as-is; falls back to the configured mode after a timeout

Regarding modes, I thought it might be useful for testing if we have a feature to also publish a fixed color for all traffic lights. (e.g., always red, always green).
Do you think it should be done by using standalone mode with 0.0 seconds for the other colors in that case, or should we have a dedicated mode for such use case.

File: simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light_publisher_node.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  const auto publish_rate = this->declare_parameter<double>("publish_rate");
+  validate_positive("publish_rate", publish_rate);
+  const auto green_duration = this->declare_parameter<double>("green_duration");
+  validate_positive("green_duration", green_duration);
+  const auto yellow_duration = this->declare_parameter<double>("yellow_duration");
+  validate_positive("yellow_duration", yellow_duration);
+  const auto red_duration = this->declare_parameter<double>("red_duration");
+  validate_positive("red_duration", red_duration);
+  const auto passthrough_timeout = this->declare_parameter<double>("passthrough_timeout");
+  validate_positive("passthrough_timeout", passthrough_timeout);
```
[the remark this answers is not in the document]
Comment (mitsudome-r) [reply simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light_publisher_node.cpp:84]: Instead of passing the node defaults, I'd prefer to update README to pass a param file for ros2 run instructions

File: simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.hpp (hunk 1 above)
Comment (takam5f2) [reply simulator/autoware_dummy_traffic_light_publisher/src/dummy_traffic_light.hpp:1]: I moved the file not for being used by public.

Thank you for feedback.

File: simulator/autoware_dummy_traffic_light_publisher/README.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+## Purpose
+
+Publish dummy traffic light signals for simulation environments where a traffic light recognition module is not available.
+
+## Modes
+
+| Mode         | Description                                                                                    |
+| ------------ | ---------------------------------------------------------------------------------------------- |
+| `standalone` | Cycles through Green -> Yellow -> Red for all traffic lights found in the vector map.          |
+| `empty`      | Publishes an empty `TrafficLightGroupArray` (no traffic light groups).                         |
```
Comment (sasakisasaki) [inline simulator/autoware_dummy_traffic_light_publisher/README.md:12]: Could you add an explanation such as sample use case for this `empty` mode: is this testing purpose use?

**Review by sasakisasaki — COMMENTED** [review]
@takam5f2 Almost reviewed and looks good! Let me ask a minor question.

File: simulator/autoware_dummy_traffic_light_publisher/README.md (hunk 6 above)
Comment (takam5f2) [reply simulator/autoware_dummy_traffic_light_publisher/README.md:12]: @sasakisasaki  
I reserved the feature to keep compatibility for the current planning simulator.  I suppose that some people don't want to test the traffic light signal functions.  However, to align interface for PSim and LSim, the topic message should be sent periodically.

**Comment by takam5f2** [comment]
@mitsudome-r  
Thanks! I added a dedicated `fixed` mode for this in 2e48436e4.

I went with a dedicated mode rather than reusing `standalone` with `0.0 s` durations.
Encoding "always green" as `green_duration > 0` and the others `= 0.0` turns `0.0` into a
magic number — the intent is hard to read from the config and raises the learning cost.

Instead:

- `mode: "fixed"` + a `fixed_color` parameter (`"green"` / `"yellow"` / `"red"`) makes the intent explicit.
- All traffic lights found in the vector map are published with the same color.

**Review by sasakisasaki — APPROVED** [review]
@takam5f2 Reviewed entire files changed. Implementation looks good, and the tests are covering possible edge cases very well. I agree with merging this PR! Good job! 👍

**Review by mitsudome-r — APPROVED** [review]
